daemon/access.go

```diff
@@ -0,0 +1,149 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/polkit"
+)
+
+type accessResult int
+
+const (
+	accessUnknown accessResult = iota
+	accessOK
+	accessUnauthorized
+	accessForbidden
+	accessCancelled
+)
+
+var polkitCheckAuthorization = polkit.CheckAuthorization
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request, deny it, or return
+// accessUnknown, which indicates the decision should be delegated to
+// the next access checker.
+type accessChecker interface {
+	canAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+// allowSnapSocket is an access checker that allows requests received
+// from the snapd-snap.socket socket.
+type allowSnapSocket struct{}
+
+func (c allowSnapSocket) canAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessOK
+	}
+	return accessUnknown
+}
+
+// denySnapSocket is an access checker that denies requests received
+// from the snapd-snap.socket socket.
+type denySnapSocket struct{}
+
+func (c denySnapSocket) canAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessUnauthorized
```

Comment:
everything else either returns OK or Unknown, but this one does this... I'll expand on this in the next comment